src/v/storage/types.h

```diff
@@ -93,7 +103,17 @@ class snapshotable_stm {
  */
 class stm_manager {
 public:
-    void add_stm(ss::shared_ptr<snapshotable_stm> stm) { _stms.push_back(stm); }
+    template<
+      typename T,
```

Comment:
no need for this to be generic, just pass in `ss::shared_ptr<compactible_stm>`

Reply:
`ss::shared_ptr` supports virtural dispatch

Reply:
Agree. Afaict the stm type hierarchy can already handle this without generics. Or create `add_compactible_stm`?

Reply:
I took another stab in force-push. We can continue discussion there.